enhancements/openshift-console-cloud-shell.md

```diff
@@ -0,0 +1,195 @@
+- Eclipse Che and the Cloud Shell are both deployed and controlled using operators but those are distinct ones.
+- Theia (or other editors) based workspaces are not supported. Only Eclipse Che workspaces with editor of type CloudShell are supported.
+
+#### Authentication and Authorization
+
+The Terminal server endpoint is exposed through a TLS route. That's the endpoint that the clients running on the browser side (xterm.js) will connect to.
+
+![](https://i.imgur.com/XWcgHeI.png)
+
+The endpoint is secured through an [OpenShift OAuth Proxy](https://github.com/openshift/oauth-proxy) sidecar. The client includes the user OpenShift token in its requests HTTP header. Only authenticated OpenShift users that satisfy the RBAC policy requirements are able to send requests to the Terminal server.
+
+In order to prevent unauthorized request to the Terminal server (coming from inside the OpenShift cluster) a [Network Policy](https://kubernetes.io/docs/concepts/services-networking/network-policies/) is used.
+
+#### Cloud Shell Pod Exec Denial Policy
+
+Users tokens are injected in the Tooling container. That's required for `oc` or `odo` to work properly. As a consequence access to the Tooling container should be allowed **only to the user of the Cloud Shell instace**. That's something that the authorization mechanism based on RBAC seen above is not able to garantee: users with editor privileges at cluster scope will be able to exec into the Pod.
+
+To prevent exec from users other then the Cloud Shell owner a [ValidatingAdmissionWebhook](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#validatingadmissionwebhook) is used.
+
+![](https://i.imgur.com/gIqTnqP.png)
+
+The Validating Webhook mechanism described above compares the user in the `AdmissionReview` request with the user that has requested the creation of the DevWorkspace CR associated to the Cloud Shell Pod. If users match the request is allowed, otherwise it's refused.
+
+This Validating Webhook mechanism requires other dynamic admission controllers to work properly:
+
+- A Mutating Admission Webhook that adds a user attribute to the DevWorkspace CR at creation time
+- A Validating Admission Webhook that block changes to the user attribute of a DevWorkspace CR
+- A Validating Admission Webhook that block changes to the Cloud Shell Pod annotation that stores the reference to the DevWorkspace CR.
+
+#### Users Token Injection in Tooling Containers
+
+That's something that we still need to figure out (see "Open Questions" section) but the injection of the user token cannot be achieved through a secret because cluster editors would be able to access it. Options are 1) xtermjs that automatically execute a command at startup 2) the `DevWorkspaces` controller that copies the token in the tooling container after it has been started.
```

Comment:
if you control the image at the other end, when you exec in, the first thing you could do is copy the user token into the pod and create a kubeconfig file and KUBECONFIG env var

Reply:
> the first thing you could do is copy the user token into the pod and create a kubeconfig file and KUBECONFIG env var

Yeah, we should be able to implement something like that.
The question is not addressed yet - where to get user's personal token.
Currently, OpenShift OAuth proxy uses SA and is not able to request user's personal token, but only some of the existing roles 
https://docs.openshift.com/container-platform/3.5/architecture/additional_concepts/authentication.html#service-accounts-as-oauth-clients

So, we need to take a look if we are able to register a dedicated OAuth client for CloudShell, OpenShift OAuth proxy could use it - our server side is able to get user's personal token and initialize kubeconfig. Will play with it shortly.

Reply:
Well, there seem to be an issue with OAuth Client for CloudShell, I've described it here https://github.com/eclipse/che/issues/16515

@deads2k Maybe you have a recommendation on how we could get user's personal token in a safe manner?
During the call, @davidfestal shared that he wondered if it's possible to have some proxy on OpenShift Console side and the URL is available for WorkspaceController...
Then CloudShell would be requested from OpenShift Console domain and we'd be able to reuse the same token OpenShift Console does.